odo/backends/avro.py

```diff
@@ -0,0 +1,412 @@
+                open(uri, 'wb'),
+                rec_writer,
+                writer_schema = sch,
+                codec=codec
+            )
+
+        return df_writer
+
+    @property
+    def writer(self):
+        if hasattr(self, '_writer'):
+            return self._writer
+        else:
+            if os.path.exists(self.path) and os.path.getsize(self.path) > 0:
+                df_writer = self._get_append_writer(self.path, self.schema, codec=self.codec)
+            else:
+                df_writer = self._get_new_writer(self.path, self.schema, codec=self.codec)
+            self._writer = df_writer
+        return df_writer
+
+    def flush(self):
+        if hasattr(self, '_writer'):
+            self._writer.close()
+            del(self._writer)
+
+@resource.register('.+\.(avro)')
+def resource_avro(uri, schema=None, codec='null', **kwargs):
+    return AVRO(uri, schema=schema, codec=codec, **kwargs)
+
+@dispatch(schema.RecordSchema)
+def discover_schema(sch):
+    return var * Record([(f.name, discover_schema(f.type)) for f in sch.fields])
+
+@dispatch(schema.UnionSchema)
+def discover_schema(sch):
+    try:
+        types = [s.type for s in sch.schemas]
+        assert "null" in types
+        types.remove("null")
+        assert len(types) == 1
+        return Option(AVRO_TYPE_MAP[types[0]])
+    except AssertionError:
+        raise TypeError("odo supports avro UnionSchema only for nullable fields." + \
+                        "Received {0}".format(str([s.type for s in sch.schemas])))
+
+@dispatch(schema.PrimitiveSchema)
+def discover_schema(sch):
+    return AVRO_TYPE_MAP[sch.type]
+
+@dispatch(schema.MapSchema)
+def discover_schema(sch):
+    # Avro map types always have string keys, see https://avro.apache.org/docs/1.7.7/spec.html#Maps
+    return Map(string, discover_schema(sch.values))
+
+@dispatch(schema.ArraySchema)
+def discover_schema(sch):
+    return var * discover_schema(sch.items)
+
+@dispatch(object)
+def discover_schema(sch):
+    raise TypeError('Unable to discover avro type %r' % type(sch).__name__)
+
+@discover.register(AVRO)
+def discover_avro(f, **kwargs):
+    return discover_schema(f.schema)
+
+def make_avsc_object(ds, name="name", namespace="default", depth=0):
+    """
+    Build Avro Schema from datashape definition
+
+    Parameters
+    ----------
+    ds : str, unicode, DataShape, or datashapes.coretypes.*
+    <name>: string -- applied to named schema elements (i.e. record, error, fixed, enum)
+    <namespace>: string -- applied to named schema elements
+    depth=0:  Tracking parameter for recursion depth.  Should not be set by user.
+
+    Examples
+    --------
+    >>> test_dshape = "var * {letter: string, value: ?int32}"
+    >>> x = make_avsc_object(test_dshape, name="my_record", namespace="com.blaze.odo")
+    >>> x == {'fields': [{'name': 'letter', 'type': 'string'},
+    ...                {'name': 'value', 'type': ['null', 'int']}],
+    ...                 'name': 'my_record',
```

Comment:
Here and below -- please clean up the indentation to make it clear that this `'name'` is a key in the outer dict.